storage/replica.go

```diff
@@ -833,20 +833,33 @@ func (r *Replica) addReadOnlyCmd(ctx context.Context, ba roachpb.BatchRequest) (
 	sp.LogEvent("command queue")
 	cmdKeys := r.beginCmds(&ba)
 
+	readOnlyLocked := false
+	defer func() {
+		// Guarantee we remove the commands from the command queue.
+		r.endCmds(cmdKeys, ba, pErr)
+
+		// Important to unlock only here to capture the timestamp cache update.
+		if readOnlyLocked {
+			r.readOnlyCmdMu.RUnlock()
+		}
+	}()
+
 	// If there are command keys (there might not be if reads are
 	// inconsistent), the read requires the leader lease.
 	if len(cmdKeys) > 0 {
-		if pErr := r.redirectOnOrAcquireLeaderLease(sp); pErr != nil {
-			r.endCmds(cmdKeys, ba, pErr)
+		if pErr = r.redirectOnOrAcquireLeaderLease(sp); pErr != nil {
```

Comment:
Don't we want to avoid acquiring the leader lease if there are no command keys (just reading the comment)? I'm guessing you wrote it though, so I'll defer to your judgement.